### DISCUSSION TITLE
Using Standard Tools or coding custom ones

### DISCUSSION THREAD

**Discussion by ahnedeee** [samuel-lucas6/Kryptor#67] [Q&A] [answered]
I wanted to know how can I achieve the same level of symmetric encryption security in Kryptor using GPG or OpenSSL?
How can I achieve authenticated encryption using them?
Or better, are there tutorials on how to implement authenticated symmetric encryption like kryptor online?

**Comment by samuel-lucas6** [comment]
I'm afraid this is rather out of scope as you're asking about other tools. A lot of detail is also missing as it's unclear whether you're talking about encrypting with a password, private key, to someone else's public key, etc. The algorithms used in Kryptor are listed [here](https://www.kryptor.co.uk/features), and I strongly suspect you cannot do something equivalent with either tool, although I've never used OpenSSL.

The most comparable tool is probably [age](https://github.com/FiloSottile/age)/[rage](https://github.com/str4d/rage). However, Kryptor has some design changes to address limitations with those. There's also [Covert](https://github.com/covert-encryption/covert), which inspired/informed some of the v4 design.

In terms of replicating Kryptor, the specification is on the [website](https://www.kryptor.co.uk/specification). [Age](https://github.com/C2SP/C2SP/blob/main/age.md) and [Covert](https://github.com/covert-encryption/covert#additional-reading) also have some.

**Comment by ahnedeee** [comment]
I’m talking about symmetric encryption using a secret key.
I also thought it might be the case that I can’t reproduce kryptor’s security using other tools.
So, how can I code the symmetric encryption using libsodium?
Are there any tutorials online?
I’m new to cryptography

**Reply by samuel-lucas6** [reply, marked as answer]
There aren't really any tutorials I can recommend. Starting out in cryptography is quite tricky and requires a lot of reading. It's best to read the libsodium [documentation](https://doc.libsodium.org/), the documentation for your libsodium binding, best/recommend practice type of guides like [this](https://github.com/samuel-lucas6/Cryptography-Guidelines) and [this](https://github.com/SalusaSecondus/CryptoGotchas), read the RFCs/Internet-Drafts for algorithms you're using like [this](https://www.rfc-editor.org/rfc/rfc8439), read up on existing protocols like the specifications I already linked, read some cryptography [books](https://github.com/samuel-lucas6/crypto-is-cool#books), and read relevant questions on [Cryptography Stack Exchange](https://crypto.stackexchange.com/) and [r/crypto](https://old.reddit.com/r/crypto/).

Some other file encryption tools to look at are [Encpipe](https://github.com/jedisct1/encpipe) and [Eureka](https://github.com/mimoo/eureka). The safest approach in libsodium is to use the [secretstream](https://doc.libsodium.org/secret-key_cryptography/secretstream) API, but not every binding supports that, and you could technically still misuse that.

**Reply by ahnedeee** [reply]
Thanks for the help

**Reply by ahnedeee** [reply]
I wanted to ask if this CLI i quickly made has any security vulnerabilities, what are some things that make it less secure:
```python

import os
import sys
import pysodium
import tempfile
import shutil

KEY_FILE = "pass.key"

def generate_key():
    return pysodium.randombytes(pysodium.crypto_aead_xchacha20poly1305_ietf_KEYBYTES)

def save_key_to_file(key):
    with open(KEY_FILE, 'wb') as key_file:
        key_file.write(key)

def load_key_from_file():
    try:
        with open(KEY_FILE, 'rb') as key_file:
            return key_file.read()
    except FileNotFoundError:
        return None

def encrypt_file(input_file, output_file, key):
    nonce = pysodium.randombytes(pysodium.crypto_aead_xchacha20poly1305_ietf_NPUBBYTES)
    with open(input_file, 'rb') as f_in, open(output_file, 'wb') as f_out:
        data = f_in.read()
        encrypted_data = pysodium.crypto_aead_xchacha20poly1305_ietf_encrypt(data, None, nonce, key)
        f_out.write(nonce + encrypted_data)

def decrypt_file(input_file, output_file, key):
    with open(input_file, 'rb') as f_in:
        data = f_in.read()
        nonce = data[:pysodium.crypto_aead_xchacha20poly1305_ietf_NPUBBYTES]
        encrypted_data = data[pysodium.crypto_aead_xchacha20poly1305_ietf_NPUBBYTES:]

        # Create a temporary file to store the decrypted data
        with tempfile.NamedTemporaryFile(delete=False) as tmp_file:
            try:
                decrypted_data = pysodium.crypto_aead_xchacha20poly1305_ietf_decrypt(encrypted_data, None, nonce, key)
                tmp_file.write(decrypted_data)
            except Exception as e:
                print(f"Decryption failed: {str(e)}")
                # If decryption fails, remove the temporary file
                os.remove(tmp_file.name)
                return

        # If decryption was successful, move the temporary file to the specified output file
        shutil.move(tmp_file.name, output_file)

def main():
    if len(sys.argv) < 4:
        print("Usage: python cli.py [encrypt/decrypt] input_file output_file")
        sys.exit(1)

    mode = sys.argv[1]
    input_file = sys.argv[2]
    output_file = sys.argv[3]

    key = load_key_from_file()

    if key is None:
        if mode == 'encrypt':
            print("No key found. Generating a new key.")
            key = generate_key()
            save_key_to_file(key)
            print(f"Encryption Key saved to {KEY_FILE}")
            # Set file permissions to 444
            os.chmod(KEY_FILE, 0o444)
        else:
            print("No key found. Unable to decrypt.")
            sys.exit(1)

    if mode == 'encrypt':
        encrypt_file(input_file, output_file, key)
        print(f"File {input_file} encrypted to {output_file}")
    elif mode == 'decrypt':
        decrypt_file(input_file, output_file, key)
        print(f"File {input_file} decrypted to {output_file}")
    else:
        print("Invalid mode. Use 'encrypt' or 'decrypt'.")
        sys.exit(1)


if __name__ == "__main__":
    main()

```

**Reply by ahnedeee** [reply]
And I also wanted to know the difference between chacha with padding fix and xchacha
what is the key commitment for?

**Reply by samuel-lucas6** [reply]
> And I also wanted to know the difference between chacha with padding fix and xchacha
> what is the key commitment for?

ChaCha20-Poly1305 with the padding fix has the same nonce size as regular ChaCha20-Poly1305, whereas XChaCha20/XChaCha20-Poly1305 has a larger nonce, making it more suitable for random generation. However, it's technically fine to randomly generate the nonce with ChaCha20-Poly1305 as long as you rotate the key frequently (e.g. every file). The padding fix is just for key commitment.

You would intuitively expect an AEAD scheme to be context/fully committing (it commits to the key, nonce, associated data, and message - all four inputs). They should really be by design. However, popular AEADs like AES-GCM and (X)ChaCha20-Poly1305 are non-committing (they commit to none of the inputs). This means tag verification can pass despite a different key being used for decryption, for example. Long story short, this causes vulnerabilities in some protocols, like with password-based encryption, if the attacker can choose parameters. There's actually a student [project](https://ethz.ch/content/dam/ethz/special-interest/infk/inst-infsec/appliedcrypto/education/theses/project_MircoStauble.pdf) on age about this.

**Reply by samuel-lucas6** [reply]
Now I don't know Python, but that looks ok from a quick glance. However, it's not typical to encrypt an entire file at once. It's recommended to encrypt the file in (e.g. 16, 32, or 64 KiB) chunks. Each chunk then gets an authentication tag after encryption. With that said, preventing chunks being rearranged/removed complicates things and is easy to get wrong. It [looks](https://github.com/stef/pysodium/issues/70) like you can use `secretstream` though, which does this for you.

Another thing is I'm not sure you need the temp file, but feel free to correct me.

**Reply by ahnedeee** [reply]
```c
#include <sodium.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define CHUNK_SIZE 4096
void generate_key(const char *key_file) {
    FILE *out = fopen(key_file, "wb");
    if (!out) {
        perror("Cannot open key file");
        exit(EXIT_FAILURE);
    }

    unsigned char key[crypto_secretstream_xchacha20poly1305_KEYBYTES];
    crypto_secretstream_xchacha20poly1305_keygen(key);

    fwrite(key, 1, sizeof key, out);
    sodium_memzero(key, sizeof key);  // Clear the memory
    fclose(out);
}

void encrypt_file(const char *input_file, const char *output_file, const unsigned char *key) {
    FILE *in = fopen(input_file, "rb");
    FILE *out = fopen(output_file, "wb");
    unsigned char buffer[CHUNK_SIZE];
    unsigned char cbuffer[CHUNK_SIZE + crypto_secretstream_xchacha20poly1305_ABYTES];
    crypto_secretstream_xchacha20poly1305_state state;
    unsigned char header[crypto_secretstream_xchacha20poly1305_HEADERBYTES];

    crypto_secretstream_xchacha20poly1305_init_push(&state, header, key);
    fwrite(header, 1, sizeof header, out);

    unsigned long long cbuf_len;
    unsigned char tag = 0;

    while (1) {
        size_t r = fread(buffer, 1, sizeof buffer, in);
        if (!r) {
            break;
        }

        // If we're on the last block of the file, set the tag to FINAL
        if (feof(in)) {
            tag = crypto_secretstream_xchacha20poly1305_TAG_FINAL;
        }

        crypto_secretstream_xchacha20poly1305_push(&state, cbuffer, &cbuf_len, buffer, r, NULL, 0, tag);
        fwrite(cbuffer, 1, cbuf_len, out);

        if (tag == crypto_secretstream_xchacha20poly1305_TAG_FINAL) {
            break;
        }
    }

    fclose(in);
    fclose(out);
}


void decrypt_file(const char *input_file, const char *output_file, const unsigned char *key) {
    FILE *in = fopen(input_file, "rb");
    FILE *out = fopen(output_file, "wb");
    unsigned char buffer[CHUNK_SIZE + crypto_secretstream_xchacha20poly1305_ABYTES];
    unsigned char dbuffer[CHUNK_SIZE];
    crypto_secretstream_xchacha20poly1305_state state;
    unsigned char header[crypto_secretstream_xchacha20poly1305_HEADERBYTES];

    fread(header, 1, sizeof header, in);
    if (crypto_secretstream_xchacha20poly1305_init_pull(&state, header, key) != 0) {
        fprintf(stderr, "Invalid header\n");
        exit(EXIT_FAILURE);
    }

    unsigned long long dbuf_len;
    unsigned char tag;
    while (1) {
        size_t r = fread(buffer, 1, sizeof buffer, in);
        if (!r) {
            break;
        }
        if (crypto_secretstream_xchacha20poly1305_pull(&state, dbuffer, &dbuf_len, &tag, buffer, r, NULL, 0) != 0) {
            fprintf(stderr, "Decryption failed\n");
            exit(EXIT_FAILURE);
        }
        fwrite(dbuffer, 1, dbuf_len, out);

        if (tag == crypto_secretstream_xchacha20poly1305_TAG_FINAL) {
            break; // Exit the loop when the final chunk is encountered
        }
    }

    fclose(in);
    fclose(out);
}
int main(int argc, char *argv[]) {
    if (argc < 3) {
        fprintf(stderr, "Usage: %s <operation> <key_file/input_password> [<input_file> <output_file>]\n", argv[0]);
        fprintf(stderr, "Operation can be: generate, encrypt, or decrypt.\n");
        return EXIT_FAILURE;
    }

    if (sodium_init() < 0) {
        // panic! the library couldn't be initialized, it is not safe to use
        fprintf(stderr, "Libsodium initialization failed\n");
        return EXIT_FAILURE;
    }

    const char *operation = argv[1];

    if (strcmp(operation, "generate") == 0) {
        if (argc != 3) {
            fprintf(stderr, "Usage for key generation: %s generate <key_file>\n", argv[0]);
            return EXIT_FAILURE;
        }
        const char *key_file = argv[2];
        generate_key(key_file);
        return EXIT_SUCCESS;
    } 
    else if (strcmp(operation, "encrypt") == 0 || strcmp(operation, "decrypt") == 0) {
        if (argc != 5) {
            fprintf(stderr, "Usage for encryption/decryption: %s <encrypt|decrypt> <key_file> <input_file> <output_file>\n", argv[0]);
            return EXIT_FAILURE;
        }

        const char *key_file = argv[2];
        unsigned char key[crypto_secretstream_xchacha20poly1305_KEYBYTES];

        FILE *kf = fopen(key_file, "rb");
        if (!kf) {
            perror("Cannot open key file");
            exit(EXIT_FAILURE);
        }
        fread(key, 1, sizeof key, kf);
        fclose(kf);

        const char *input_file = argv[3];
        const char *output_file = argv[4];

        if (strcmp(operation, "encrypt") == 0) {
            encrypt_file(input_file, output_file, key);
        } 
        else if (strcmp(operation, "decrypt") == 0) {
            decrypt_file(input_file, output_file, key);
        } 
        else {
            fprintf(stderr, "Invalid operation\n");
            return EXIT_FAILURE;
        }

        sodium_memzero(key, sizeof key);  // Clear the memory after usage
    } 
    else {
        fprintf(stderr, "Invalid operation\n");
        return EXIT_FAILURE;
    }

    return EXIT_SUCCESS;
}

```

Is this code better?

**Comment by ahnedeee** [comment]
> > And I also wanted to know the difference between chacha with padding fix and xchacha
> > what is the key commitment for?
> 
> ChaCha20-Poly1305 with the padding fix has the same nonce size as regular ChaCha20-Poly1305, whereas XChaCha20/XChaCha20-Poly1305 has a larger nonce, making it more suitable for random generation. However, it's technically fine to randomly generate the nonce with ChaCha20-Poly1305 as long as you rotate the key frequently (e.g. every file). The padding fix is just for key commitment.
> 
> You would intuitively expect an AEAD scheme to be context/fully committing (it commits to the key, nonce, associated data, and message - all four inputs). They should really be by design. However, popular AEADs like AES-GCM and (X)ChaCha20-Poly1305 are non-committing (they commit to none of the inputs). This means tag verification can pass despite a different key being used for decryption, for example. Long story short, this causes vulnerabilities in some protocols, like with password-based encryption, if the attacker can choose parameters. There's actually a student [project](https://ethz.ch/content/dam/ethz/special-interest/infk/inst-infsec/appliedcrypto/education/theses/project_MircoStauble.pdf) on age about this.

so is using the algoithm in the code I provided [above](https://github.com/samuel-lucas6/Kryptor/discussions/67#discussioncomment-7216747) safe?

**Reply by samuel-lucas6** [reply]
I'm afraid I can't keep peer reviewing your code, especially since I don't program in those languages. As to your question, secretstream isn't committing because it uses (X)ChaCha20-Poly1305. However, if you're only ever using your tool offline on your machine, you should be fine.

**Reply by ahnedeee** [reply]
In terms of peer reviewing my code, don't worry.
In terms of committing, when should it be necessary?

**Reply by samuel-lucas6** [reply]
All future AEAD schemes should be fully committing ideally because that's what you'd intuitively expect and to prevent potential vulnerabilities. The vulnerabilities require attacker control over inputs, meaning it's an online concern, like if there's a server that knows the encryption key derived from a password and performs decryption for you (an oracle). Another example would be sending an encrypted file to someone and it decrypting to different plaintexts for different recipients.

**Comment by ahnedeee** [comment]
By the way, does kryptor "encrypt then mac" or no?
And if it doesn't, is it fully committing?

**Reply by samuel-lucas6** [reply]
This is the last question I'll answer in this thread. It used to, but I switched to ChaCha20-Poly1305 for better efficiency on ARM, although that's technically Encrypt-then-MAC internally. The padding fix only provides key commitment. In the next file format or another tool, I will use another AEAD or a generic composition AEAD.
